### PR TITLE
add support for action groups

### PR DESCRIPTION
**Pull request by with-heart**
Per discussion in #3554, this PR adds support for action groups.

This allows us to execute named (and possibly nested) groups of actions together:

```ts
const machine = createMachine(
  {
    entry: 'group',
  },
  {
    actions: {
      group: ['action1', 'group2'],
      group2: ['action2']
      action1: () => console.log('action1'),
      action2: () => console.log('action2'),
    },
  },
)

interpret(machine).start()

// "action1", "action2"
```

### CODE REVIEW

**Comment by with-heart** [comment]
I should probably add the docs in a separate PR so we can add a version `Badge`

File: packages/core/test/actions.test.ts [hunk 1]
```diff
@@ -1682,6 +1682,32 @@ describe('purely defined actions', () => {
   });
 });
 
+describe('action groups', () => {
+  it('should execute actions provided as type strings in a group', () => {
```
Comment (davidkpiano) [inline packages/core/test/actions.test.ts:1686]: Awesome! Can you also add tests to support:

- Mixture of strings, objects, functions
- An action group that references another action that references that action group (edge-case)
- Some more edge-cases I can't think of right now

File: .changeset/wicked-pandas-add.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      group: ['action1', 'action2'],
+      action1: () => console.log('action 1'),
+      action2: () => console.log('action 2')
+    }
+  }
+);
+
+const service = interpret(machine).start();
+
+service.send(service, 'event'); // "action 1", "action 2"
```
Comment (Andarist) [inline .changeset/wicked-pandas-add.md:25]:
```suggestion
service.send({ type: 'event' }); // logs: "action 1" and "action 2"
```

File: docs/guides/actions.md [hunk 3]
```diff
@@ -212,6 +212,57 @@ entry: send({ type: 'SOME_EVENT' });
...
+      group: ['action1', 'action2'],
+      action1: () => console.log('action 1'),
+      action2: () => console.log('action 2')
+    }
+  }
+);
+
+const service = interpret(machine).start();
+
+service.send(service, 'event'); // "action 1", "action 2"
```
Comment (Andarist) [inline docs/guides/actions.md:237]:
```suggestion
service.send({ type: 'event' }); // logs: "action 1" and "action 2"
```

File: packages/core/src/actionTypes.ts [hunk 4]
```diff
@@ -19,3 +19,4 @@ export const error = ActionTypes.ErrorCustom;
 export const update = ActionTypes.Update;
 export const choose = ActionTypes.Choose;
 export const pure = ActionTypes.Pure;
+export const actionGroup = ActionTypes.ActionGroup;
```
Comment (Andarist) [inline packages/core/src/actionTypes.ts:22]: I wouldn't introduce this one. I feel like arrays are already pretty descriptive when it comes to such groups and we can use them directly.

File: packages/core/src/types.ts [hunk 5]
```diff
@@ -788,7 +790,8 @@ type MachineOptionsActions<
         TContext,
         Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>,
         Cast<Prop<TIndexedActions, K>, BaseActionObject>
-      >;
+      >
+    | Array<Cast<Prop<TIndexedActions, K>, BaseActionObject>['type']>;
```
Comment (Andarist) [inline packages/core/src/types.ts:794]: We probably should accept more things here. For instance those should be valid:
```
group: [assign(), 'another']
group: [assign(), sendParent()]
group: [assign(), { type: 'another' }]
```

Comment (with-heart) [reply packages/core/src/types.ts:794]: I'm working on shifting it to:

```ts
{
  [K in keyof TEventsCausingActions]?:
    | ActionObject<
        TContext,
        Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>
      >
    | ActionFunction<
        TContext,
        Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>,
        Cast<Prop<TIndexedActions, K>, BaseActionObject>
      >
    | Array<
        | ActionObject<
            TContext,
            Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>
          >
        | ActionFunction<
            TContext,
            Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>,
            Cast<Prop<TIndexedActions, K>, BaseActionObject>
          >
        | Cast<Prop<TIndexedActions, K>, BaseActionObject>['type']
      >;
}
```

It'll accept objects and functions though it still leaves us with the same type problems as `pure`. At least they're consistent 😆

Comment (with-heart) [reply packages/core/src/types.ts:794]: Also thinking I should change `Cast<Prop<TIndexedActions, K>, BaseActionObject>['type']` to `keyof TEventsCausingActions`

- autocompletion of `type` strings
- the typegen portion of this automatically adds `type` strings in groups to `eventsCausingActions`, so it wouldn't do anything for you on the first use of a `type`, but if you used it in later groups you'd get autocomplete

Think there'd be any downsides to that?

File: packages/core/src/actionTypes.ts (hunk 4 above)
Comment (with-heart) [reply packages/core/src/actionTypes.ts:22]: I added this following @davidkpiano's comment about action groups being syntactic sugar for `xstate.actionGroup` over here: https://github.com/statelyai/xstate/discussions/3554#discussioncomment-3499861

If we didn't use an `ActionObject` to describe these, would we just flatten the group arrays in `toActionObjects`?

File: packages/core/test/actions.test.ts (hunk 1 above)
Comment (with-heart) [reply packages/core/test/actions.test.ts:1686]: Added some tests for groups of each type as well as a mixture

> An action group that references another action that references that action group (edge-case)

Would we just test that this throws a maximum call stack exceeded error for the time being? 

Was playing around with a way to track and prevent cyclical action calls last night because `pure` and `choose` can suffer from the same problem, but it feels pretty thorny

**Comment by with-heart** [comment]
Updated the example in docs/changeset to show a mix of `type` strings, objects, and functions

File: packages/core/test/actions.test.ts (hunk 1 above)
Comment (davidkpiano) [reply packages/core/test/actions.test.ts:1686]: > Would we just test that this throws a maximum call stack exceeded error for the time being?

While creating the action, you can have some variable in closure (like a `Set`) that keeps track of actions already added to the group, and if it encounters it again (`set.has(someActionType)`), you can detect that cycle.

Comment (with-heart) [reply packages/core/test/actions.test.ts:1686]: Just want to make sure I'm understanding. Are you thinking this `Set` would be within a cycle test or in the actual codebase?

**Comment by davidkpiano** [comment]
I still love the idea of this; what is needed to get it through?

**Comment by with-heart** [comment]
@davidkpiano hoping to shift back to working on `xstate` stuff this week. This time with a digital garden so I can document all the things as I do them!

Need to spend some time going through this again and remembering where things left off 😆

**Comment by davidkpiano** [comment]
@with-heart Can we re-target this for v5? (`next` branch)

**Comment by with-heart** [comment]
> @with-heart Can we re-target this for v5? (`next` branch)

Yeeee, I'll get conflicts resolved and make sure it's working again

**Comment by with-heart** [comment]
It was easier to just reimplement this from scratch than deal with the rebase lol. Will reopen from a new branch shortly
